speechbrain/lobes/models/huggingface_transformers/llama.py

```diff
@@ -0,0 +1,135 @@
+    https://huggingface.co/transformers/installation.html
+
+    Arguments
+    ---------
+    source : str
+        HuggingFace hub name: e.g "meta-llama/Llama-2-7b-chat-hf"
+    save_path : str
+        Path (dir) of the downloaded model.
+    bnb_config : transformers.BitsAndBytesConfig
+        BitsAndBytesConfig enabling quantisation of the model. If not specified, the model weights will be loaded with weight_precision_load dtype.
+    freeze : bool (default: false)
+        If True, the model is frozen. If False, the model will be trained
+        alongside with the rest of the pipeline.
+    pad_token : str (default: "PAD")
+        String representation of the padding token. This may change from one model to another.
+    padding_side : str (default: "right")
+        One of right or left.
+    weight_precision_load : torch.dtype (default: torch.bfloat16)
+        If no bnb_config is given, this parameter defines the loading type of the parameters of the model. This is useful to reduce memory footprint, but it does not change the compute dtype. For this just refer to mixed precision training in SpeechBrain.
+    **kwargs
+        Extra keyword arguments passed to the `from_pretrained` function. This can be used, for instance, to change the type of attention. The HuggingFace documentation gives the full list of parameters which may be model dependent.
+
+    Example
+    -------
+    >>> model_hub = "meta-llama/Llama-2-7b-chat-hf"
+    >>> save_path = "savedir"
+    >>> model = LLaMA(model_hub, save_path) # doctest: +SKIP
+    >>> tokens = torch.tensor([[1, 1]])
+    >>> attention_mask = torch.tensor([[1, 1]])
+    >>> outputs = model(tokens, attention_mask) # doctest: +SKIP
+    """
+
+    def __init__(
+        self,
+        source: str,
+        save_path: str,
+        bnb_config: BitsAndBytesConfig = None,
+        freeze: bool = False,
+        pad_token: str = "[PAD]",
+        padding_side: Literal["left", "right"] = "right",
+        weight_precision_load: torch.dtype = torch.bfloat16,
+        **kwargs,
+    ) -> None:
+        self.pad_token = pad_token
+        self.padding_side = padding_side
+        self.source = source
+        self.save_path = save_path
+        self.bnb_config = bnb_config
+
+        if self.bnb_config is not None:
+            logger.info(
+                "LlaMA will be quantised following the given configuration."
+            )
+
+        super().__init__(
+            source=source,
+            save_path=save_path,
+            freeze=freeze,
+            with_casual_lm=True,
+            quantization_config=self.bnb_config,
+            torch_dtype=weight_precision_load,
+            **kwargs,
+        )
+
+        self.load_tokenizer(source=source, pad_token=self.pad_token)
+
+        # We resize the token embeddings size to a factor of 8 to maximise
+        # the use of tensorcores.
+        self.model.resize_token_embeddings(
+            len(self.tokenizer), pad_to_multiple_of=8
+        )
```

Comment:
I still tend to think that this should be part of `huggingface.py` but if you really think its better here then ok but i dont see valid reasons to not move it to the abstract class